src/data_tools/src/jsf_data.cpp

```diff
@@ -11,21 +11,271 @@
+    }
+    stringstream time_ss;
+    time_ss << data_time;
+    ping.time_string_ = time_ss.str();
+    boost::posix_time::time_duration const diff = data_time - epoch;
+    ping.time_stamp_ = diff.total_milliseconds();
 
+ 
+    jsf_msg_header jsf_hdr_next;
+    input.read(reinterpret_cast<char*>(&jsf_hdr_next),sizeof(jsf_hdr_next));
+
+    while((jsf_hdr_next.msg_type!=SONAR_DATA_TYPE)&&(!input.eof())){
+        skip_data(input, jsf_hdr_next);
+        input.read(reinterpret_cast<char*>(&jsf_hdr_next),sizeof(jsf_hdr_next));
+
+        if (jsf_hdr_next.start_marker != SONAR_MESSAGE_HEADER_START){
+            cout << "Invalid file format! start marker: "<<jsf_hdr_next.start_marker << endl;
+            exit(0);
+        }
+    }
+    
+    // continue to read next channel
+    jsf_sonar_data_msg_header jsf_sonar_data_hdr_next;    
+    input.read(reinterpret_cast<char*>(&jsf_sonar_data_hdr_next),sizeof(jsf_sonar_data_hdr_next));
+    jsf_sss_ping_side ping_side_nxt;
+    if (jsf_hdr_next.start_marker == SONAR_MESSAGE_HEADER_START){
+        ping_side_nxt = process_side_scan_ping_side(input, jsf_hdr_next, jsf_sonar_data_hdr_next);
+    }  
+
+        if(jsf_hdr.channel_num==0){
+            ping.port = ping_side_cur;
+            ping.stbd = ping_side_nxt;
+        }
+        else{
+            ping.stbd = ping_side_cur;
+            ping.port = ping_side_nxt;
+        }
+  
+
+    // determine if it is a bad ping 
+    if(((jsf_hdr.channel_num+jsf_hdr_next.channel_num)!=1)||((jsf_sonar_data_hdr.ping_time_in_sec!=jsf_sonar_data_hdr_next.ping_time_in_sec))){
+        ping.is_bad = true;
+    }
+    else{
+        ping.is_bad = false;
+    }
+    ping.first_in_file_ = false;
+    return ping;
+
+}
+
+
+
+
+
+jsf_sss_ping::PingsT read_jsf_file(ifstream& input, jsf_msg_header& jsf_hdr){
+    jsf_sss_ping::PingsT pings;
+    jsf_sonar_data_msg_header jsf_sonar_data_hdr;
+    while (!input.eof()){
+        input.read(reinterpret_cast<char*>(&jsf_hdr),sizeof(jsf_hdr));
+        if (jsf_hdr.start_marker != SONAR_MESSAGE_HEADER_START){
+            cout << "Invalid file format! start marker: "<<jsf_hdr.start_marker << endl;
+            break;
+        }
+
+        if (jsf_hdr.msg_type==SONAR_DATA_TYPE){
+            jsf_sss_ping ping= process_side_scan_ping(input, jsf_hdr, jsf_sonar_data_hdr);
+
+            if (ping.is_bad){
+                cout << pings.size()+1 <<"th ping is a bad ping, ignore!"<<endl;
+            }
+            else{
+                pings.push_back(ping);
+            }
+            
+        }
+        else{
+            skip_data(input,jsf_hdr);
+        }
+    }
+    cout<<"Read Done!"<<endl;
+
+    if (!pings.empty()) {
+        pings[0].first_in_file_ = true;
+    }
+    return pings;
+
+}
+
+
+template <typename ReturnType>
```

Comment:
You do not need a template here as far as I can tell, please remove it.

Reply:
Or if you want to make it generic over different kinds of types, use it in the return type and inside the function. But then you would need to make `read_jsf_file` generic as well.